src/cargo/core/compiler/standard_lib.rs

```diff
@@ -147,41 +149,58 @@ pub fn generate_std_roots(
     let std_pkgs = package_set.get_many(std_ids)?;
     // Generate a map of Units for each kind requested.
     let mut ret = HashMap::new();
-    for pkg in std_pkgs {
-        let lib = pkg
-            .targets()
-            .iter()
-            .find(|t| t.is_lib())
-            .expect("std has a lib");
-        let unit_for = UnitFor::new_normal();
-        // I don't think we need to bother with Check here, the difference
-        // in time is minimal, and the difference in caching is
-        // significant.
-        let mode = CompileMode::Build;
-        let features = std_features.activated_features(pkg.package_id(), FeaturesFor::NormalOrDev);
 
+    let std_pkg_infos: Vec<_> = std_pkgs
+        .iter()
+        .map(|pkg| {
+            let lib = pkg
+                .targets()
+                .iter()
+                .find(|t| t.is_lib())
+                .expect("std has a lib");
+            let unit_for = UnitFor::new_normal();
+            let features =
+                std_features.activated_features(pkg.package_id(), FeaturesFor::NormalOrDev);
+            (pkg, lib, unit_for, features)
+        })
+        .collect();
+
+    for kinds in package_set
+        .packages()
+        .map(|pkg| pkg.explicit_kinds(requested_kinds, explicit_host_kind))
```

Comment:
This loop might look a bit nicer with less indentation if `flat_map` is used here instead of `map` perhaps?